File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {

    int[] visited;
    ArrayList<Integer> result;
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<List<Integer>>();

        for(int i = 0 ;i < numCourses; i++){
            graph.add(new ArrayList<Integer>());
        }

        for(int[] pair: prerequisites){
            graph.get(pair[1]).add(pair[0]);
        }

        visited = new int[numCourses];
        for(int i = 0; i < numCourses; i++){
            if(visited[i] == 0 && checkCycle(i, graph)){
                return new int[0];
            }
        }

        visited = new int[numCourses];
        result = new ArrayList<Integer>();
        for(int i = 0; i < numCourses; i++){
            if(visited[i] == 0){
                dfs(i, graph);
            }
        }

        int[] path = new int[numCourses];
        for(int i = 0; i < numCourses; i++){
            path[i] = result.get(result.size() - i - 1);
        }
        return path;
    }

    public boolean checkCycle(int node, List<List<Integer>> graph){

        visited[node] = 1;
        boolean isCycle = false;
        for(int nb: graph.get(node)){
            if(visited[nb] == 1){
                return true;
            }

            if(visited[nb] == 0){
                isCycle = isCycle || checkCycle(nb, graph);
            }
        }

        visited[node] = 2;
        return isCycle;
    }

    public void dfs(int node, List<List<Integer>> graph){

        visited[node] = 1;

        for(int el: graph.get(node)){
            if(visited[el] == 0){
                dfs(el, graph);
            }
        }

        result.add(node);
    }
}

File: 210-course-schedule-ii/course-schedule-ii.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """

        adj = []
        inc = []
        for el in range(0, numCourses):
            adj.append([])
            inc.append(0)

        for pre in prerequisites:
            a = pre[0];
            b = pre[1];
            inc[a]+=1
            adj[b].append(a)
        
        visited = []
        for el in range(0, numCourses):
            visited.append(0)

        for el in range(0, numCourses):
            if visited[el] == 0 and self.checkCycle(el, adj, visited):
                return []

        stack = []
        for node in range(0, numCourses):
            if inc[node] == 0:
                stack.append(node)

        ans = []
        while len(stack) > 0:
            node = stack.pop()
            ans.append(node)
            for child in adj[node]:
                inc[child]-=1

                if inc[child] == 0:
                    stack.append(child) 

        # print(ans)
        return ans

    def checkCycle(self, node, adj, visited):
        
        visited[node] = 1

        flag = False
        for child in adj[node]:
            if visited[child] == 1:
                return True
            
            if visited[child] == 0:
                visited[child] = 1
                flag = flag or self.checkCycle(child, adj, visited)
                if flag:
                    return True
        
        visited[node] = 2
        return False
        